src/actions/dropbox/dropbox.ts

```diff
@@ -0,0 +1,175 @@
+import {AwsKms} from "../../crypto/aws_kms"
+import * as Hub from "../../hub"
+
+export class DropboxAction extends Hub.OAuthAction {
+    name = "dropbox"
+    label = "Dropbox"
+    iconName = "dropbox/dropbox.png"
+    description = "Send query results directly to a file in your Dropbox."
+    supportedActionTypes = [Hub.ActionType.Query, Hub.ActionType.Dashboard]
+    usesStreaming = false
+    minimumSupportedLookerVersion = "6.2.0"
+    requiredFields = []
+    params = [
+      {
+        description: "Dropbox Application Key",
+        label: "Application Key",
+        name: "appKey",
+        required: true,
+        sensitive: true,
+      },
+      {
+        description: "Dropbox Secret Key",
+        label: "Secret Key",
+        name: "secretKey",
+        required: true,
+        sensitive: true,
+      },
+    ]
+
+  async execute(request: Hub.ActionRequest) {
+    const filename = request.formParams.filename
+    const directory = request.formParams.directory
+    const ext = request.attachment!.fileExtension
+
+    const drop = this.dropboxClientFromRequest(request)
+    const resp = new Hub.ActionResponse()
+    if (request.attachment && request.attachment.dataBuffer) {
+      const fileBuf = request.attachment.dataBuffer
+      await drop.filesUpload({path: `/${directory}/${filename}.${ext}`, contents: fileBuf}).then((_dropResp) => {
+        resp.success = true
+        resp.state = new Hub.ActionState()
+        resp.state.data = "reset"
+      }).catch((err: any) => {
+        winston.error(`Upload unsuccessful: ${JSON.stringify(err)}`)
+        resp.success = false
+        resp.state = new Hub.ActionState()
+        resp.state.data = "reset"
+      })
+    } else {
+      resp.success = false
+      resp.message = "No data sent from Looker to be sent to Dropbox"
+    }
+    return resp
+  }
+
+  async form(request: Hub.ActionRequest) {
+    const form = new Hub.ActionForm()
+    form.fields = []
+
+    const drop = this.dropboxClientFromRequest(request)
+    return new Promise<Hub.ActionForm>((resolve, reject) => {
+      drop.filesListFolder({path: ""})
+        .then((resp) => {
+          form.fields = [{
+            description: "Dropbox directory where file will be saved",
+            label: "Save in",
+            name: "directory",
+            options: resp.entries.map((entries) => ({name: entries.name, label: entries.name})),
+            required: true,
+            type: "select",
+          }, {
+            label: "Filename",
+            name: "filename",
+            type: "string",
+          }]
+          resolve(form)
+        })
+        .catch((_error: DropboxTypes.Error<DropboxTypes.files.ListFolderError>) => {
+          winston.info("Could not list Dropbox folders")
+          const creds = JSON.stringify({app: request.params.appKey, secret: request.params.secretKey})
+          const kms = new AwsKms()
+          kms.encrypt(creds).then((ciphertextBlob: string) => {
```

Comment:
I don't think that individual actions should be manually leveraging particular encryption strategies. Can we move the encryption into an interface and have individual actions leverage encryption through that interface?